## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd
import numpy as np

ecommerce_df = pd.read_csv('Resources/client_dataset.csv')
ecommerce_df.head(10)

first      last                                           job  \
0     Donald   Harding                                  Immunologist   
1    Tiffany     Myers                               Music therapist   
2    Shannon    Watson                                  Immunologist   
3     Nathan     Baker                         Accounting technician   
4  Christina  Schwartz                                  Chiropractor   
5    Jessica     Reyes  Production designer, theatre/television/film   
6      Brian     Moore                                       Barista   
7      Jesus    Wright                            Theme park manager   
8    Rebecca     Brown                                   Chiropodist   
9       Greg    Garcia                            Secretary, company   

                 phone                                email  client_id  \
0   793-904-7725x39308     harding.donald.7185@sullivan.com      58515   
1     201.442.4543x942    myers.t.6537@ferguson-johnson.net      37609   
2    687.737.9424x8503                swatson8146@payne.net      57113   
3     827-788-8123x012               bakernathan@benson.com      46554   
4         265-829-3643  christinaschwartz9252@mcconnell.com      92089   
5           5016915923             jessica.reyes@grimes.net      33615   
6  +1-893-400-2733x982          moore.official@martinez.com      87213   
7   605.947.9467x66513        wright.jesus.7579@meadows.com      46709   
8     665-955-0156x926                 rbrown850@lewis.info      73362   
9   (231)342-7216x1421                  ggarcia@bentley.biz      58789   

   order_id  order_date  order_week  order_year         item_id     category  \
0   8953482  2023-04-28          17        2023  EUD29711-63-6U        decor   
1   8069089  2023-05-19          20        2023  XDA18116-89-4A  consumables   
2   1902144  2023-01-29           4        2023  ABE59463-05-7E     software   
3   9031802  2023-04-25          17        2023  ZMM00836-65-0C  consumables   
4   1322274  2023-05-28          21        2023  BZX55559-12-3X  consumables   
5   1327730  2023-05-02          18        2023  BMB23158-88-9A        decor   
6   8080280  2023-04-19          16        2023  UAC50632-25-4E  consumables   
7   8946782  2023-02-14           7        2023  UEU25316-01-1U  consumables   
8   9394118  2023-05-13          19        2023  DDA06248-98-7Z  consumables   
9   9665852  2023-04-14          15        2023  DUU85042-87-5E    furniture   

          subcategory  unit_price  unit_cost  unit_weight  qty  line_number  
0            wall art     1096.80     762.71         7.50  105            1  
1                pens       24.95      15.09         1.49   21            0  
2  project management       13.52       7.86         1.68   39            6  
3                pens       36.42      24.85         1.23   29            3  
4                misc      195.10     108.17        46.43   20            1  
5              plants      172.84     154.44         2.79   78            0  
6    kitchen supplies      245.56     231.95         1.73   56            5  
7                pens       20.99      16.97         5.12   22            4  
8      paper products       83.86      59.00         0.27   49            3  
9             seating       66.88      54.15         2.47   78            1

In [2]:
# View the column names in the data
list(ecommerce_df.columns)

['first',
 'last',
 'job',
 'phone',
 'email',
 'client_id',
 'order_id',
 'order_date',
 'order_week',
 'order_year',
 'item_id',
 'category',
 'subcategory',
 'unit_price',
 'unit_cost',
 'unit_weight',
 'qty',
 'line_number']

In [3]:
# Use the describe function to gather some basic statistics
ecommerce_df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.
ecommerce_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [5]:
# What three item categories had the most entries?
top_three_categories = ecommerce_df['category'].value_counts().head(3)
top_three_categories

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [6]:
# For the category with the most entries, which subcategory had the most entries?
# Save the top category as a new variable
top_category = ecommerce_df['category'].value_counts().idxmax()

print(f"The top category is {top_category}.")

# Filter the dataframe to show rows from the top category only
filtered_df = ecommerce_df[ecommerce_df['category'] == top_category]

# Use value_counts on the sub category column within this filtered dataframe
top_sub_category = filtered_df['subcategory'].value_counts().idxmax()

print(f"The top sub category within {top_category} is {top_sub_category}.")



The top category is consumables.
The top sub category within consumables is bathroom supplies.


In [7]:
# Which five clients had the most entries in the data?
client_counts = ecommerce_df['client_id'].value_counts()

top_five_clients = client_counts.head(5)
top_five_clients

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [8]:
# Store the client ids of those top 5 clients in a list.
top_five_clients_list = top_five_clients.index.tolist()

print(top_five_clients_list)

[33615, 66037, 46820, 38378, 24741]


In [9]:
# How many total units (the qty column) did the client with the most entries order?
client_33615 = ecommerce_df[ecommerce_df['client_id'] == 33615]

total_units_33615 = client_33615['qty'].sum()
print(total_units_33615)



64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [10]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
ecommerce_df['sub_total'] = ecommerce_df['unit_price'] * ecommerce_df['qty']

sub_total_ecommerce_df = ecommerce_df[['client_id', 'order_id', 'unit_price', 'qty', 'sub_total']]
sub_total_ecommerce_df.head(10)

client_id  order_id  unit_price  qty  sub_total
0      58515   8953482     1096.80  105  115164.00
1      37609   8069089       24.95   21     523.95
2      57113   1902144       13.52   39     527.28
3      46554   9031802       36.42   29    1056.18
4      92089   1322274      195.10   20    3902.00
5      33615   1327730      172.84   78   13481.52
6      87213   8080280      245.56   56   13751.36
7      46709   8946782       20.99   22     461.78
8      73362   9394118       83.86   49    4109.14
9      58789   9665852       66.88   78    5216.64

In [11]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.

def calculate_shipping_price(unit_weight, qty):
    order_weight = unit_weight * qty
    if order_weight > 50:
        return 7 * order_weight
    else:
        return 10 * order_weight
    
ecommerce_df['shipping_price'] = ecommerce_df.apply(lambda row: calculate_shipping_price(row['unit_weight'], row['qty']), axis=1)

shipping_price_df = ecommerce_df[['client_id', 'order_id', 'unit_price', 'unit_weight', 'qty', 'sub_total', 'shipping_price']]
shipping_price_df.head(10)



client_id  order_id  unit_price  unit_weight  qty  sub_total  \
0      58515   8953482     1096.80         7.50  105  115164.00   
1      37609   8069089       24.95         1.49   21     523.95   
2      57113   1902144       13.52         1.68   39     527.28   
3      46554   9031802       36.42         1.23   29    1056.18   
4      92089   1322274      195.10        46.43   20    3902.00   
5      33615   1327730      172.84         2.79   78   13481.52   
6      87213   8080280      245.56         1.73   56   13751.36   
7      46709   8946782       20.99         5.12   22     461.78   
8      73362   9394118       83.86         0.27   49    4109.14   
9      58789   9665852       66.88         2.47   78    5216.64   

   shipping_price  
0         5512.50  
1          312.90  
2          458.64  
3          356.70  
4         6500.20  
5         1523.34  
6          678.16  
7          788.48  
8          132.30  
9         1348.62

In [12]:
# Create a column for sales tax using the subtotal, shipping price and a 9.25% tax rate
ecommerce_df['sales_tax'] = (((ecommerce_df['sub_total']) + (ecommerce_df['shipping_price'])) * 0.0925).round(2)

sales_tax_df = ecommerce_df[['client_id', 'order_id', 'unit_price', 'unit_weight', 'qty', 'sub_total', 'shipping_price', 'sales_tax']]
sales_tax_df.head(10)

client_id  order_id  unit_price  unit_weight  qty  sub_total  \
0      58515   8953482     1096.80         7.50  105  115164.00   
1      37609   8069089       24.95         1.49   21     523.95   
2      57113   1902144       13.52         1.68   39     527.28   
3      46554   9031802       36.42         1.23   29    1056.18   
4      92089   1322274      195.10        46.43   20    3902.00   
5      33615   1327730      172.84         2.79   78   13481.52   
6      87213   8080280      245.56         1.73   56   13751.36   
7      46709   8946782       20.99         5.12   22     461.78   
8      73362   9394118       83.86         0.27   49    4109.14   
9      58789   9665852       66.88         2.47   78    5216.64   

   shipping_price  sales_tax  
0         5512.50   11162.58  
1          312.90      77.41  
2          458.64      91.20  
3          356.70     130.69  
4         6500.20     962.20  
5         1523.34    1387.95  
6          678.16    1334.73  
7          788.48     115.65  
8          132.30     392.33  
9         1348.62     607.29

In [13]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
ecommerce_df['total_price'] = (ecommerce_df['sub_total'] + ecommerce_df['shipping_price'] + ecommerce_df['sales_tax'])

total_price_df = ecommerce_df[['client_id', 'order_id', 'unit_price', 'unit_weight', 'qty', 'sub_total', 'shipping_price', 'sales_tax', 'total_price']]
total_price_df.head(10)

client_id  order_id  unit_price  unit_weight  qty  sub_total  \
0      58515   8953482     1096.80         7.50  105  115164.00   
1      37609   8069089       24.95         1.49   21     523.95   
2      57113   1902144       13.52         1.68   39     527.28   
3      46554   9031802       36.42         1.23   29    1056.18   
4      92089   1322274      195.10        46.43   20    3902.00   
5      33615   1327730      172.84         2.79   78   13481.52   
6      87213   8080280      245.56         1.73   56   13751.36   
7      46709   8946782       20.99         5.12   22     461.78   
8      73362   9394118       83.86         0.27   49    4109.14   
9      58789   9665852       66.88         2.47   78    5216.64   

   shipping_price  sales_tax  total_price  
0         5512.50   11162.58    131839.08  
1          312.90      77.41       914.26  
2          458.64      91.20      1077.12  
3          356.70     130.69      1543.57  
4         6500.20     962.20     11364.40  
5         1523.34    1387.95     16392.81  
6          678.16    1334.73     15764.25  
7          788.48     115.65      1365.91  
8          132.30     392.33      4633.77  
9         1348.62     607.29      7172.55

In [14]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
ecommerce_df['total_cost'] = ((ecommerce_df['unit_cost'] * ecommerce_df['qty'])) + ecommerce_df['shipping_price']

total_cost_df = ecommerce_df[['client_id', 'order_id', 'unit_price', 'unit_cost', 'unit_weight', 'qty', 'sub_total', 'shipping_price', 'sales_tax', 'total_price', 'total_cost']]
total_cost_df.head(10)


client_id  order_id  unit_price  unit_cost  unit_weight  qty  sub_total  \
0      58515   8953482     1096.80     762.71         7.50  105  115164.00   
1      37609   8069089       24.95      15.09         1.49   21     523.95   
2      57113   1902144       13.52       7.86         1.68   39     527.28   
3      46554   9031802       36.42      24.85         1.23   29    1056.18   
4      92089   1322274      195.10     108.17        46.43   20    3902.00   
5      33615   1327730      172.84     154.44         2.79   78   13481.52   
6      87213   8080280      245.56     231.95         1.73   56   13751.36   
7      46709   8946782       20.99      16.97         5.12   22     461.78   
8      73362   9394118       83.86      59.00         0.27   49    4109.14   
9      58789   9665852       66.88      54.15         2.47   78    5216.64   

   shipping_price  sales_tax  total_price  total_cost  
0         5512.50   11162.58    131839.08    85597.05  
1          312.90      77.41       914.26      629.79  
2          458.64      91.20      1077.12      765.18  
3          356.70     130.69      1543.57     1077.35  
4         6500.20     962.20     11364.40     8663.60  
5         1523.34    1387.95     16392.81    13569.66  
6          678.16    1334.73     15764.25    13667.36  
7          788.48     115.65      1365.91     1161.82  
8          132.30     392.33      4633.77     3023.30  
9         1348.62     607.29      7172.55     5572.32

In [15]:
# Create a column for the profit of each line using line cost and line price
ecommerce_df['profit'] = (ecommerce_df['total_price'] - ecommerce_df['total_cost']).round(2)

total_profit_df = ecommerce_df[['client_id', 'order_id', 'unit_price', 'unit_cost', 'unit_weight', 'qty', 'sub_total', 'shipping_price', 'sales_tax', 'total_price', 'total_cost', 'profit']]
total_profit_df.head(10)


client_id  order_id  unit_price  unit_cost  unit_weight  qty  sub_total  \
0      58515   8953482     1096.80     762.71         7.50  105  115164.00   
1      37609   8069089       24.95      15.09         1.49   21     523.95   
2      57113   1902144       13.52       7.86         1.68   39     527.28   
3      46554   9031802       36.42      24.85         1.23   29    1056.18   
4      92089   1322274      195.10     108.17        46.43   20    3902.00   
5      33615   1327730      172.84     154.44         2.79   78   13481.52   
6      87213   8080280      245.56     231.95         1.73   56   13751.36   
7      46709   8946782       20.99      16.97         5.12   22     461.78   
8      73362   9394118       83.86      59.00         0.27   49    4109.14   
9      58789   9665852       66.88      54.15         2.47   78    5216.64   

   shipping_price  sales_tax  total_price  total_cost    profit  
0         5512.50   11162.58    131839.08    85597.05  46242.03  
1          312.90      77.41       914.26      629.79    284.47  
2          458.64      91.20      1077.12      765.18    311.94  
3          356.70     130.69      1543.57     1077.35    466.22  
4         6500.20     962.20     11364.40     8663.60   2700.80  
5         1523.34    1387.95     16392.81    13569.66   2823.15  
6          678.16    1334.73     15764.25    13667.36   2096.89  
7          788.48     115.65      1365.91     1161.82    204.09  
8          132.30     392.33      4633.77     3023.30   1610.47  
9         1348.62     607.29      7172.55     5572.32   1600.23

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [16]:
# Check your work using the totals above
order_id_2742071 = ecommerce_df.loc[ecommerce_df['order_id'] == 2742071, ['order_id', 'total_price']]

order_id_2742071_sum = order_id_2742071['total_price'].sum().round(2)
print(f"Order ID 2742071 had a total price of ${order_id_2742071_sum:,}")
print("-------------------------------------------------------------")

order_id_2173913 = ecommerce_df.loc[ecommerce_df['order_id'] == 2173913, ['order_id', 'total_price']]

order_id_2173913_sum = order_id_2173913['total_price'].sum().round(2)
print(f"Order ID 2173913 had a total price of ${order_id_2173913_sum:,}")
print("-------------------------------------------------------------")

order_id_6128929 = ecommerce_df.loc[ecommerce_df['order_id'] == 6128929, ['order_id', 'total_price']]

order_id_6128929_sum = order_id_6128929['total_price'].sum().round(2)
print(f"Order ID 6128929 had a total price of ${order_id_6128929_sum:,}")
print("-------------------------------------------------------------")



Order ID 2742071 had a total price of $152,811.89
-------------------------------------------------------------
Order ID 2173913 had a total price of $162,388.71
-------------------------------------------------------------
Order ID 6128929 had a total price of $923,441.25
-------------------------------------------------------------


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [17]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
for client in top_five_clients_list:
    total_spend = ecommerce_df[ecommerce_df['client_id'] == client]['total_price'].sum()
    print(f"Client {client} spent: ${total_spend:,.2f}")

    
        
        


Client 33615 spent: $8,377,308.52
Client 66037 spent: $10,259,514.79
Client 46820 spent: $9,743,794.36
Client 38378 spent: $12,906,550.87
Client 24741 spent: $82,268,892.02


In [18]:
# Create a summary DataFrame showing the totals for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 

# Initialize an empty list to store the summary data for each top client
summary_data = []

# Iterate over each client in the top five clients list
for client in top_five_clients_list:
    # Filter the DataFrame for the current client
    client_data = ecommerce_df[ecommerce_df['client_id'] == client]
    
    # Sum the relevant columns
    qty = client_data['qty'].sum()
    shipping_price = client_data['shipping_price'].sum()
    total_price = client_data['total_price'].sum()
    total_cost = client_data['total_cost'].sum()
    total_profit = client_data['profit'].sum()
    
    # Append the results as a dictionary
    summary_data.append({
        'client_id': client,
        'Units': qty,
        'Total Shipping Price': shipping_price,
        'Total Revenue': total_price,
        'Total Cost': total_cost,
        'Total Profit': total_profit
    })

# Create a new summary DataFrame from the list of dictionaries
top_five_summary_df = pd.DataFrame(summary_data)

# Display the summary DataFrame
top_five_summary_df

                                                                                                                                                            


client_id   Units  Total Shipping Price  Total Revenue   Total Cost  \
0      33615   64313            1828984.89     8377308.52   6175313.91   
1      66037   43018            1395151.85    10259514.79   7004482.98   
2      46820   75768            1601448.84     9743794.36   7007191.64   
3      38378   73667            3429455.40    12906550.87   9634720.98   
4      24741  239862            5126448.37    82268892.02  45688899.71   

   Total Profit  
0    2201994.61  
1    3255031.81  
2    2736602.72  
3    3271829.89  
4   36579992.31

In [19]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 
money_columns = ['Total Shipping Price', 'Total Revenue', 'Total Cost', 'Total Profit']

# Define a function that converts a dollar amount to millions.
def currency_format_millions(dollar_amount):
    return dollar_amount / 1000000

# Apply the currency_format_millions function to only the money columns. 
top_five_summary_df[money_columns] = top_five_summary_df[money_columns].apply(currency_format_millions)

# Rename the columns to reflect the change in the money format. 
top_five_summary_df.rename(columns={
    'client_id': 'Client ID',
    'Total Shipping Price': 'Shipping (Millions)',
    'Total Revenue': 'Total Revenue (Millions)',
    'Total Cost': 'Total Cost (Millions)',
    'Total Profit': 'Total Profit (Millions)',
    
}, inplace=True)

# Display the updated data frame.
top_five_in_millions_df = top_five_summary_df[['Client ID', 'Shipping (Millions)', 'Total Revenue (Millions)', 'Total Cost (Millions)', 'Total Profit (Millions)']]
top_five_in_millions_df



Client ID  Shipping (Millions)  Total Revenue (Millions)  \
0      33615             1.828985                  8.377309   
1      66037             1.395152                 10.259515   
2      46820             1.601449                  9.743794   
3      38378             3.429455                 12.906551   
4      24741             5.126448                 82.268892   

   Total Cost (Millions)  Total Profit (Millions)  
0               6.175314                 2.201995  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
3               9.634721                 3.271830  
4              45.688900                36.579992

In [20]:
# Sort the updated data by "Total Profit (millions)" from highest to lowest and assign the sort to a new DatFrame.
sorted_df = top_five_in_millions_df.sort_values(by='Total Profit (Millions)', ascending=False)

sorted_df

Client ID  Shipping (Millions)  Total Revenue (Millions)  \
4      24741             5.126448                 82.268892   
3      38378             3.429455                 12.906551   
1      66037             1.395152                 10.259515   
2      46820             1.601449                  9.743794   
0      33615             1.828985                  8.377309   

   Total Cost (Millions)  Total Profit (Millions)  
4              45.688900                36.579992  
3               9.634721                 3.271830  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
0               6.175314                 2.201995

In [21]:
"""
Client 24741 has generated the most revenue and profit for this business. However I believe there
may be a mistake in the expected solutions to these exercises. 
Sales tax should be deducted from total profit because sales tax is not profit for the business. 
All businesses act as agents of the state when collecting sales tax. Sales tax must be remitted
to the tax authorities and therefore should not be calculated as profit.

This mistake, if I am correct could result in a 6.96 million dollar accounting error for Client 24741 alone.
"""


'\nClient 24741 has generated the most revenue and profit for this business. However I believe there\nmay be a mistake in the expected solutions to these exercises. \nSales tax should be deducted from total profit because sales tax is not profit for the business. \nAll businesses act as agents of the state when collecting sales tax. Sales tax must be remitted\nto the tax authorities and therefore should not be calculated as profit.\n\nThis mistake, if I am correct could result in a 6.96 million dollar accounting error for Client 24741 alone.\n'